src/activity_queue.rs

```diff
@@ -220,27 +201,145 @@ pub(crate) fn generate_request_headers(inbox_url: &Url) -> HeaderMap {
     headers
 }
 
+/// A simple activity queue which spawns tokio workers to send out requests
+/// When creating a queue, it will spawn a task per worker thread
+/// Uses an unbounded mpsc queue for communication (i.e, all messages are in memory)
+pub(crate) struct ActivityQueue {
+    // Our "background" tasks
+    senders: Vec<UnboundedSender<SendActivityTask>>,
+    // Round robin of the sender list
+    last_sender_idx: AtomicUsize,
+    // Stats shared between the queue and workers
+    stats: Arc<Stats>,
+}
+
+/// Simple stat counter to show where we're up to with sending messages
+/// This is a lock-free way to share things between tasks
+/// When reading these values it's possible (but extremely unlikely) to get stale data if a worker task is in the middle of transitioning
+#[derive(Default)]
+struct Stats {
+    pending: AtomicUsize,
+    running: AtomicUsize,
+    dead: AtomicUsize,
+    complete: AtomicUsize,
+}
+
+/// We need to retry activity sending in case the target instances is temporarily unreachable.
+/// In this case, the task is stored and resent when the instance is hopefully back up. This
+/// list shows the retry intervals, and which events of the target instance can be covered:
+/// - 60s (one minute, service restart)
+/// - 60min (one hour, instance maintenance)
+/// - 60h (2.5 days, major incident with rebuild from backup)
+/// TODO: make the intervals configurable
+const MAX_RETRIES: usize = 3;
+const BACKOFF: usize = 60;
+
+/// A tokio spawned worker which is responsible for submitting requests to federated servers
+async fn worker(
+    client: ClientWithMiddleware,
+    timeout: Duration,
+    mut receiver: UnboundedReceiver<SendActivityTask>,
+    stats: Arc<Stats>,
+) {
+    while let Some(message) = receiver.recv().await {
+        // Update our counters as we're now "running" and not "pending"
+        stats.pending.fetch_sub(1, Ordering::Relaxed);
+        stats.running.fetch_add(1, Ordering::Relaxed);
+
+        // This will use the retry helper method below, with an exponential backoff
+        // If the task is sleeping, tokio will use work-stealing to keep it busy with something else
+        let outcome = retry(|| do_send(&message, &client, timeout), MAX_RETRIES, BACKOFF).await;
+
+        // "Running" has finished, check the outcome
+        stats.running.fetch_sub(1, Ordering::Relaxed);
+
+        match outcome {
+            Ok(_) => {
+                stats.complete.fetch_add(1, Ordering::Relaxed);
+            }
+            // We might want to do something here
+            Err(_err) => {
+                stats.dead.fetch_add(1, Ordering::Relaxed);
+            }
+        }
+    }
+}
+
+impl ActivityQueue {
+    fn new(client: ClientWithMiddleware, timeout: Duration, worker_count: usize) -> Self {
+        // Keep a vec of senders to send our messages to
+        let mut senders = Vec::with_capacity(worker_count);
```

Comment:
This is actually something I was wondering about recently. At the moment we have lots of new users on Lemmy, and instance admins need to increase this worker count manually to keep federation synchronized. However the workers seem very lightweight, lemmy.ml is running fine with 200k running at once.

So maybe you can tell me if this fixed worker count is really necessary, or if we can remove it entirely and spawn workers on demand. This would also require that unused workers get cleaned up automatically when the queue empties. 

Related to this, Im wondering how tokio workers compare to previous actix workers, would there be any difference in cpu or memory usage?

Reply:
So regarding the original implementation I don't know how concurrent it would be, considering that there isn't really any work stealing going on between threads.

I have another PR for lemmy to use tokio instead of actix rt to bring up the server, as I feel (& definitely need to test this with benchmarking!) that we should be using tokio's work stealing instead of actix rt.  However it panicked when I got rid of actix rt because `background-jobs` uses it, hence this PR.

The reason why I think this is the case is that [actix rt brings up multiple single threaded executors](https://docs.rs/actix-rt/latest/actix_rt/), but they work independently and do not share tasks with other threads:

>To achieve similar performance to multi-threaded, work-stealing runtimes, applications using actix-rt will create multiple, mostly disconnected, single-threaded runtimes. This approach has good performance characteristics for workloads where the majority of tasks have similar runtime expense.

This is great at pinning tcp sockets to specific threads and saving on context switching, but the brunt of the work lemmy is doing is sharing DB connections and this background activity stuff, highly variable in terms of workload.   This is why I think it's probably much more suited to a work stealing queue, especially if we are sleeping/retrying a lot of the time.

I will preface this that I think I'll need to work on making a bit of a benchmark to prove this theory out as it's hard to know without some reference whether this will make things faster.  It might make things slower!

I am not sure how concurrent the existing implementation is, given that it's running on individual single threaded executors & while you might submit 200k tasks there are only a fixed number of workers running them concurrently.  For actix rt with default I believe it's one single-threaded runtime per thread.

In terms of workers on demand: yes we can possibly look at scaling up/down dynamically, but I'm thinking a good first step is to have a number of workers as a multiplier as number of threads available.  Like 128 * [`available_parallelism()`](https://doc.rust-lang.org/std/thread/fn.available_parallelism.html).   Once again, benchmarks appear to be necessary here.  We might also need to make the reqwest connection pool match this amount.

The [tokio documentation](https://tokio.rs/tokio/tutorial/spawning) states that there is about 64 bytes of memory per task.  I believe that these task are going to live on the stack now which might mean we need to box them if there are going to be millions.

So the related question is solved by needing to benchmark both approaches

Reply:
I am completely ignorant when it comes to different async runtimes, it sounds like you understand it much better. as a reference, the highest I saw was around 200.000 workers on lemmy.ml, running on an 8cpu vps, with the lemmy backend taking roughly one cpu core. Postgres took everything else, and ram was not a problem. So optimizing this is not necessary yet, but cant hurt to make it more effective. For benchmarking we could deploy something to lemmy.ml, as long as it wont crash. 

@asonix Maybe you also have some thoughts on this.

Reply:
> So maybe you can tell me if this fixed worker count is really necessary, or if we can remove it entirely and spawn workers on demand

> In terms of workers on demand: yes we can possibly look at scaling up/down dynamically, but I'm thinking a good first step is to have a number of workers as a multiplier as number of threads available

I think @Nutomic is right here.. 

@cetra3 's code uses `tokio::spawn` to create a set of 64 (worker_count) tokio tasks and then uses round-robin to distribute the incoming tasks to those 64 workers. This looks very much like you're building a load-leveling system to distribute the load.

But I don't think this approach makes a lot of sense on top of tokio: Tokio already has an internal [pool of worker threads](https://github.com/tokio-rs/tokio/discussions/3858#discussioncomment-869878) that it distributes the load to. That queue has a fixed number of worker threads (by default = cpu_count).

So now you basically have a queue of 64 "fake" workers (green threads) that are scheduled on-demand on 8 real tokio worker threads, that each have their own unbounded queue of tasks to run. It seems like you're adding a second level of indirection that I don't think gives you any advantages. Having more "fake" workers on top of the cpu_count real workers shouldn't improve performance any either. 

Instead I think it would make more sense to remove the round-robin and just rely on the internal tokio work queue to make the scheduling decisions. The work stealing happens on the layer of the real tokio threads, so by adding that second layer of green threads you're actually preventing work stealing between each of the 64 green threads. The reason is that you're making fixed assignments of each thing to one of the 64 threads instead of letting tokio decide where to put it.

I'd use a single `mpsc` queue with a single loop that reads from the mpsc queue and directly `tokio::spawn`s a task for each thing. That way this code becomes a fair bit simpler and (imo) should also be faster. [Just like the example in the main tokio docs docs](https://tokio.rs/tokio/tutorial/spawning).

Reply:
If the number of tasks should be limited I think something like this should be done instead: https://github.com/tokio-rs/tokio/discussions/2648#discussioncomment-34885

That way the rest of the tokio runtime is not affected if there's too many pending tasks. But then the mpsc queue should also be bounded and work has to be dropped based on some criteria (e.g. prioritize posts over comments, comments over votes).

Reply:
Hey there! Just commenting since I was tagged: I think this is a fine change to make. Especially because lemmy hasn't had persisted storage for background tasks. I might recommend taking a look at https://github.com/rafaelcaricio/backie if you want to grow from this inline queue to a more structured and persisted queue in the future.

I'll also say that using 200K workers with background-jobs probably wasn't a great idea, but if it worked for you then it worked lol.